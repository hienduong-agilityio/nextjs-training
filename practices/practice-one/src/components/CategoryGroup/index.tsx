'use client';

// Libraries
import { usePathname, useSearchParams } from 'next/navigation';
import { twMerge } from 'tailwind-merge';

// Components
import Link from 'next/link';

// Helpers
import { capitalizeText } from '@/helpers';

// Constants
import { ALL_CATEGORIES, SEARCH_PARAMS } from '@/constants';

interface ICategoryGroupProps {
  title: string;
  items: { name: string; count: number }[];
}

export function CategoryGroup({ title, items }: ICategoryGroupProps) {
  const pathname = usePathname();
  const searchParams = useSearchParams();

  const currentCategory =
    searchParams.get(SEARCH_PARAMS.CATEGORY) ?? ALL_CATEGORIES.ALL;

  const handleCategoryClick = (itemName: string) => {
    const updatedParams = new URLSearchParams();

    // Update the category query parameter
    if (itemName !== ALL_CATEGORIES.ALL) {
      updatedParams.set(SEARCH_PARAMS.CATEGORY, itemName);
    }

    return `${pathname}?${updatedParams.toString()}`;
  };

  return (
    <div className="mb-6 bg-secondary-100 p-4 rounded">
      <h3 className="text-lg font-bold mb-4">{title}</h3>
      <ul className="space-y-2 text-sm text-gray-700">
        {items.map((item) => {
          const isActive =
            item.name === ALL_CATEGORIES.ALL
              ? currentCategory === ALL_CATEGORIES.ALL || !currentCategory
              : item.name === currentCategory;

          return (
            <li className="flex justify-between" key={item.name}>
              <Link
                href={handleCategoryClick(item.name)}
                className={twMerge(
                  'w-full block font-semibold py-2 rounded hover:bg-secondary-200',
                  isActive
                    ? 'bg-primary-500 text-primary-200'
                    : 'text-gray-800',
                )}
              >
                {capitalizeText(item.name)}
              </Link>
            </li>
          );
        })}
      </ul>
    </div>
  );
}
