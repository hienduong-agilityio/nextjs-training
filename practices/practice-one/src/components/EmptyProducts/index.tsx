import Link from 'next/link';

export const EmptyProducts = ({
  message,
  actionLabel,
  actionHref,
}: {
  message: string;
  actionLabel: string;
  actionHref: string;
}) => (
  <div className="flex flex-col items-center gap-4 py-16">
    <p className="text-xl text-gray-600">{message}</p>
    <Link
      href={actionHref}
      className="inline-block px-6 py-2 text-white bg-blue-500 rounded-lg"
    >
      {actionLabel}
    </Link>
  </div>
);
