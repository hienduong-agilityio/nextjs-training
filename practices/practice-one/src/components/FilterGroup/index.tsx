interface FilterGroupProps {
  title: string;
  items: { name: string; count: number }[];
}

export const FilterGroup = ({ title, items }: FilterGroupProps) => {
  return (
    <div className="mb-6 bg-secondary-100 p-4 rounded">
      <h3 className="text-lg font-bold mb-4">{title}</h3>
      <ul className="space-y-2 text-sm text-gray-700">
        {items.map((item) => (
          <li key={item.name} className="flex justify-between py-2">
            <button className="font-semibold hover:text-primary-200 ">
              {item.name}
            </button>
            <span>({item.count})</span>
          </li>
        ))}
      </ul>
    </div>
  );
};
