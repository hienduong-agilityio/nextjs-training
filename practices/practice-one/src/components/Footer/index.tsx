// Icon
import {
  FacebookIcon,
  LogoIcon,
  MasterCardIcon,
  PaypalIcon,
  TwitterIcon,
  VisaIcon,
  WesternIcon,
} from '@/icons';

// Mock
import { FOOTER_NAVIGATION } from '@/mocks';

// Components
import Link from 'next/link';

// Constants
import { ROUTE } from '@/constants';

export const Footer = () => {
  return (
    <footer className="w-full bg-primary-50">
      <div className="md:container px-4 pt-6 md:mx-auto sm:px-6 md:px-8 lg:px-12 xl:px-16 sm:pt-8 md:pt-12 lg:pt-16 xl:pt-20">
        <div className="py-8">
          {/* Main Footer Content */}
          <div className="flex flex-col gap-10 lg:flex-row lg:justify-between">
            {/* Logo Section */}
            <div className="max-w-[290px] flex flex-col gap-4">
              <Link href={ROUTE.PRODUCTS} className="flex items-center gap-2">
                <LogoIcon aria-label="logo" />
                <span className="text-lg font-bold sm:text-xl lg:text-2xl">
                  E-Comm
                </span>
              </Link>
              <span className="text-sm font-normal leading-normal">
                Lorem Ipsum is simply dummy text of the printing and typesetting
                industry. Lorem Ipsum has been the industry standard dummy text
                ever. Since the 1500s, when an unknown printer.
              </span>
            </div>

            <div className="flex flex-col w-full gap-8 sm:flex-row sm:justify-evenly lg:w-3/4">
              {/* Follow Us */}
              <div className="lg:max-w-[220px] flex flex-col gap-4">
                <h4 className="font-medium text-foreground-400">Follow Us</h4>
                <span className="text-sm">
                  Since the 1500s, when an unknown printer took a galley of type
                  and scrambled.
                </span>
                <div className="flex gap-4">
                  <Link href="https://www.facebook.com/" aria-label="Facebook">
                    <FacebookIcon color="#385C8E" />
                  </Link>
                  <Link href="https://www.twitter.com/" aria-label="Twitter">
                    <TwitterIcon />
                  </Link>
                </div>
              </div>

              {/* Contact Us */}
              <div className="lg:max-w-[150px] flex flex-col gap-4">
                <h4 className="font-medium text-foreground-400">Contact Us</h4>
                <address className="text-sm not-italic">
                  E-Comm, 4578 Marmora Road, Glasgow D04 89GR
                </address>
              </div>
            </div>
          </div>

          {/* Footer Links */}
          <div className="grid grid-cols-1 gap-8 pt-16 sm:grid-cols-2 lg:grid-cols-4">
            {FOOTER_NAVIGATION.map((item) => (
              <ul key={item.title}>
                <li>
                  <h4 className="mb-5 font-medium text-foreground-400">
                    {item.title}
                  </h4>
                  <ul>
                    {item.items.map((item) => (
                      <li
                        key={item.name}
                        className="mb-3 text-sm text-content2 hover:text-primary-700"
                      >
                        <Link href={item.url} className="hover:underline">
                          {item.name}
                        </Link>
                      </li>
                    ))}
                  </ul>
                </li>
              </ul>
            ))}
          </div>
        </div>

        {/* Footer Bottom */}
        <div className="flex flex-col items-center justify-between gap-6 py-6 border-t-2 border-secondary-100 lg:flex-row ">
          <p className="text-sm text-secondary-500">
            &#169; 2018 Ecommerce theme by www.bisenbaev.com
          </p>
          <div className="flex gap-2">
            <WesternIcon aria-label="Western" />
            <MasterCardIcon aria-label="MasterCard" />
            <PaypalIcon aria-label="Paypal" />
            <VisaIcon aria-label="Visa" />
          </div>
        </div>
      </div>
    </footer>
  );
};
