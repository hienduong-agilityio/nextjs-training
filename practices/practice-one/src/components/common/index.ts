export { default as Button } from './Button';
export { default as InputField } from './InputField';
export { default as Tabs } from './Tabs';
export { default as Select } from './Select';
export { default as IconButton } from './IconButton';
export * from './IconButton';
