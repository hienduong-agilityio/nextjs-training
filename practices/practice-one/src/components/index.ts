export * from './Header';
export { default as SearchBox } from './SearchBox';
export * from './SearchBox';
export * from './ProductTab';

// Common
export * from './common';
