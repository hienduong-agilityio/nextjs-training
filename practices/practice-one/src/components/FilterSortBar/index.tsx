'use client';

// Hooks
import { useSearchParams, useRouter } from 'next/navigation';
import { useEffect, useState } from 'react';

// Components
import { Select } from '@/components';

// Constants
import { SEARCH_PARAMS, SORT_PRODUCT_OPTIONS } from '@/constants';

interface FilterSortBarProps {
  itemCount: number;
  sortOptions: string[];
  showOptions: string[];
}

export const FilterSortBar = ({
  itemCount = 0,
  sortOptions = SORT_PRODUCT_OPTIONS,
  showOptions = ['6'],
}: FilterSortBarProps) => {
  const searchParams = useSearchParams();
  const router = useRouter();

  const [selectedLimit, setSelectedLimit] = useState<string>(showOptions[0]);
  const [selectedSort, setSelectedSort] = useState<string>(sortOptions[0]);

  // TODO: Remove useEffect.
  useEffect(() => {
    // Get query parameters from the URL
    const limitParam = searchParams.get(SEARCH_PARAMS.LIMIT) ?? showOptions[0];
    const sortParam = searchParams.get(SEARCH_PARAMS.SORT_BY) ?? sortOptions[0];

    setSelectedLimit(limitParam);
    setSelectedSort(sortParam);
  }, [searchParams, showOptions, sortOptions]);

  const updateQueryParams = (param: string, value: string) => {
    const currentParams = new URLSearchParams(searchParams.toString());
    currentParams.set(param, value);

    // Reset pagination if a filter changes
    if (param === SEARCH_PARAMS.LIMIT) {
      currentParams.delete(SEARCH_PARAMS.PAGE);
    }

    router.push(`?${currentParams.toString()}`);
  };

  const handleShowChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const value = event.target.value;

    setSelectedLimit(value);
    updateQueryParams(SEARCH_PARAMS.LIMIT, value);
  };

  const handleSortChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const value = event.target.value;

    setSelectedSort(value);
    updateQueryParams(SEARCH_PARAMS.SORT_BY, value);
  };

  return (
    <div className="bg-secondary-100 p-4 shadow-md md:sticky rounded top-0 z-30 w-full">
      <div className="flex flex-col sm:flex-row sm:justify-between sm:items-center space-y-4 sm:space-y-0">
        {/* Item count */}
        <div className="text-sm w-max text-gray-600">{itemCount} Items</div>

        {/* Filters */}
        <div className="flex flex-col sm:flex-row sm:items-center sm:space-x-4 space-y-4 sm:space-y-0">
          <div className="flex gap-2 items-center">
            <span>Sort</span>
            {/* TODO: Custom style for select arrow by tailwinds classes*/}
            <Select
              options={sortOptions}
              value={selectedSort}
              customClass="w-auto bg-secondary-100 capitalize"
              onChange={handleSortChange}
            />
          </div>
          <div className="flex gap-2 items-center">
            <span>Show</span>
            <Select
              options={showOptions}
              value={selectedLimit}
              customClass="w-auto bg-secondary-100"
              onChange={handleShowChange}
            />
          </div>
        </div>
      </div>
    </div>
  );
};
