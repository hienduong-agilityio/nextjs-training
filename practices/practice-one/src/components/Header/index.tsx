// Icon
import { CartIcon, LogoIcon, ProfileIcon } from '@/icons';

// Components
import { SearchBox } from '@/components';

interface IHeaderProps {
  cartItemCount?: number;
}

export const Header = ({ cartItemCount = 0 }: IHeaderProps) => {
  return (
    <header className="container flex flex-wrap items-center w-full px-4 py-4 pb-6 mx-auto md:gap-8 md:justify-start xl:justify-between lg:pb-12 sm:px-6 md:px-8 lg:px-12 xl:px-16 sm:py-6 md:py-8 lg:py-10 xl:py-12">
      <div className="flex flex-wrap gap-5 xl:justify-between lg:w-auto">
        <a href="/" className="flex items-center gap-2">
          {/* Logo */}
          <LogoIcon aria-label="logo" />
          <span className="text-lg font-bold sm:text-xl lg:text-2xl">
            E-Comm
          </span>
        </a>

        {/* Search Box */}
        <SearchBox
          customClass={{
            container: 'md:justify-end',
            input: 'w-full md:w-[400px] lg:w-[520px]',
            inputContainer: 'border-primary-300',
            button: 'bg-primary-300 text-white',
          }}
        />
      </div>

      <div className="flex items-center gap-4 mt-5 sm:gap-6 lg:mt-0">
        {/* Profile */}
        <div className="flex items-center gap-2 text-sm sm:text-base lg:text-lg">
          <ProfileIcon
            aria-label="profile"
            className="flex items-center"
            size={20}
          />
          <span>My account</span>
        </div>

        {/* Cart */}
        <div className="relative flex items-center">
          <CartIcon aria-label="cart" size={25} itemCount={cartItemCount} />
        </div>
      </div>
    </header>
  );
};
