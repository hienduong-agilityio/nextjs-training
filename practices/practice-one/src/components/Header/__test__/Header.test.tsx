// Libraries
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { useRouter, useSearchParams } from 'next/navigation';

// Components
import { Header } from '@/components';

// Mock the Next.js router and searchParams
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
  useSearchParams: jest.fn(),
}));

describe('Header Component - UI Tests', () => {
  beforeEach(() => {
    // Mock `useRouter` behavior
    (useRouter as jest.Mock).mockReturnValue({
      push: jest.fn(),
    });

    // Mock `useSearchParams` behavior
    (useSearchParams as jest.Mock).mockReturnValue(new URLSearchParams(''));
  });

  it('renders the header with default props and matches snapshot', () => {
    const { asFragment } = render(<Header />);

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the logo with correct text', () => {
    render(<Header />);

    const logo = screen.getByLabelText('logo');
    expect(logo).toBeInTheDocument();
    expect(screen.getByText('E-Comm')).toBeInTheDocument();
  });

  it('renders the profile section', () => {
    render(<Header />);

    const profileIcon = screen.getByLabelText('profile');
    expect(profileIcon).toBeInTheDocument();
    expect(screen.getByText('My account')).toBeInTheDocument();
  });

  it('renders the cart section', () => {
    render(<Header />);

    const cartIcon = screen.getByLabelText('cart');
    expect(cartIcon).toBeInTheDocument();
  });

  it('renders the search box', () => {
    render(<Header />);

    const searchInput = screen.getByRole('textbox');
    expect(searchInput).toBeInTheDocument();

    const searchButton = screen.getByRole('button', { name: /Search/i });
    expect(searchButton).toBeInTheDocument();
  });

  it('renders the cart item count if provided', () => {
    const cartItemCount = 3;
    render(<Header cartItemCount={cartItemCount} />);

    expect(screen.getByText(cartItemCount.toString())).toBeInTheDocument();
  });
});
