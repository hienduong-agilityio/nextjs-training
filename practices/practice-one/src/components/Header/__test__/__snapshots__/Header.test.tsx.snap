// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component - UI Tests renders the header with default props and matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="flex items-center justify-between w-full gap-6 px-4 py-4 md:container md:mt-10 md:mx-auto sm:px-6 md:px-8 lg:px-12 xl:px-16"
  >
    <a
      class="flex items-center gap-2"
      href="/"
    >
      <svg
        aria-label="logo"
        fill="none"
        height="44.217"
        viewBox="0 0 44 44.217"
        width="44"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#40BFFF"
          height="44.217"
          rx="16"
          transform="scale(1)"
          width="44"
        />
        <path
          clip-rule="evenodd"
          d="M31.062 20.372c0.954 0.959 0.954 2.514 0 3.474l-7.334 7.37c-0.954 0.959-2.502 0.959-3.456 0l-7.334-7.37a2.465 2.465 0 0 1 0-3.474l7.334-7.37a2.436 2.436 0 0 1 3.457 0l7.333 7.37ZM22 18.213l-3.876 3.896L22 26.004l3.876-3.895L22 18.213Z"
          fill="#fff"
          fill-rule="evenodd"
        />
      </svg>
      <span
        class="w-max text-lg font-bold sm:text-xl lg:text-2xl"
      >
        E-Comm
      </span>
    </a>
    <div
      class="w-full flex items-center justify-end lg:justify-between gap-4"
    >
      <div
        class="hidden lg:block"
      >
        <div
          class="flex md:justify-end"
        >
          <div
            class="w-full"
          >
            <div
              class="flex items-center gap-1 border rounded-md focus-within:ring-2 focus-within:ring-primary-100 lg:rounded-r-none p-2 border-primary-300"
            >
              <span>
                <svg
                  class="lg:hidden block"
                  fill="none"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8.875 16.75C13.2242 16.75 16.75 13.2242 16.75 8.875C16.75 4.52576 13.2242 1 8.875 1C4.52576 1 1 4.52576 1 8.875C1 13.2242 4.52576 16.75 8.875 16.75Z"
                    stroke="#40BFFF"
                    stroke-miterlimit="10"
                    stroke-width="2"
                  />
                  <path
                    d="M14.5 14.5L19 19"
                    stroke="#40BFFF"
                    stroke-linecap="round"
                    stroke-miterlimit="10"
                    stroke-width="2"
                  />
                </svg>
              </span>
              <input
                class="flex-1 outline-none px-3 py-2 bg-transparent h-16 text-gray-700 appearance-none [&::-webkit-search-cancel-button]:cursor-pointer w-full lg:w-[400px]"
                id="gsearch"
                name="gsearch"
                placeholder="Enter your query..."
                type="search"
                value=""
              />
            </div>
          </div>
          <button
            class="items-center gap-2 rounded p-2 bg-opacity-100 shadow-md hover:bg-primary-400 disabled:bg-primary-100 px-7 rounded-l-none hidden lg:block bg-primary-300 text-white"
            disabled=""
            type="submit"
          >
            Search
          </button>
        </div>
      </div>
      <div>
        <button
          class="block lg:hidden cursor-pointer"
          type="button"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 20 20"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.875 16.75C13.2242 16.75 16.75 13.2242 16.75 8.875C16.75 4.52576 13.2242 1 8.875 1C4.52576 1 1 4.52576 1 8.875C1 13.2242 4.52576 16.75 8.875 16.75Z"
              stroke="#9098B1"
              stroke-miterlimit="10"
              stroke-width="2"
            />
            <path
              d="M14.5 14.5L19 19"
              stroke="#9098B1"
              stroke-linecap="round"
              stroke-miterlimit="10"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <div
        class="flex items-center gap-2"
      >
        <div
          class="flex items-center gap-2"
        >
          <svg
            aria-label="profile"
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M19.356 16.2615A9.856 9.856 0 0 0 13.949 10.64a5.948 5.948 0 0 0 2.1715 -4.546a6.172 6.172 0 0 0 -1.6515 -4.105A6.1625 6.1625 0 0 0 10.495 0.0465c-0.4975 -0.062 -1.0005 -0.062 -1.498 0a6.1625 6.1625 0 0 0 -3.97 1.9435a6.172 6.172 0 0 0 -1.65 4.1035a5.949 5.949 0 0 0 2.1715 4.546A9.8455 9.8455 0 0 0 0.14 16.2615a2.354 2.354 0 0 0 0.33 2.2135A2.5885 2.5885 0 0 0 2.5505 19.5h14.405a2.5825 2.5825 0 0 0 2.078 -1.025a2.345 2.345 0 0 0 0.3225 -2.2135Zm-13.43 -10.168a3.7115 3.7115 0 0 1 1.0125 -2.48a3.7055 3.7055 0 0 1 2.417 -1.1515a3.215 3.215 0 0 1 0.79 0a3.663 3.663 0 0 1 2.9675 2.6145a3.67 3.67 0 0 1 -1.133 3.7905a3.663 3.663 0 0 1 -6.0535 -2.7735ZM2.5485 17.0625a7.766 7.766 0 0 1 2.855 -3.5415a7.7565 7.7565 0 0 1 11.55 3.5415H2.5485Z"
              fill="#22262A"
              fill-rule="evenodd"
            />
          </svg>
          <span
            class="hidden text-sm w-max md:text-base lg:text-lg lg:block"
          >
            My account
          </span>
          <div
            class="relative flex items-center"
          >
            <svg
              aria-label="cart"
              fill="none"
              height="25"
              viewBox="0 0 40 32"
              width="25"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 28c.447 0 .81-.354.81-.79a.801.801 0 0 0-.81-.792.801.801 0 0 0-.81.791c0 .438.362.792.81.792ZM24 28c.447 0 .81-.354.81-.79a.801.801 0 0 0-.81-.792.801.801 0 0 0-.81.791c0 .438.362.792.81.792Z"
                fill="#22262A"
                stroke="#22262A"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-miterlimit="10"
                stroke-width="2.0833333333333335"
              />
              <path
                d="M4 4h3l3 18h16l3-14H10"
                stroke="#22262A"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2.0833333333333335"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
